src/lib.rs

```diff
@@ -1697,6 +1719,56 @@ pub struct Keystore<
     task_scope: AsyncScope<'a, ()>,
 }
 
+#[derive(Clone)]
+pub struct EncryptingResourceAdapter<T> {
+    cipher: Cipher<ChaChaRng>,
+    _phantom: std::marker::PhantomData<T>,
+}
+
+impl<T> EncryptingResourceAdapter<T> {
+    fn new(key: KeyTree) -> Self {
+        Self {
+            cipher: Cipher::new(key, ChaChaRng::from_entropy()),
+            _phantom: Default::default(),
+        }
+    }
+
+    fn cast<S>(&self) -> EncryptingResourceAdapter<S> {
+        EncryptingResourceAdapter {
+            cipher: self.cipher.clone(),
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<T: Serialize + DeserializeOwned> LoadStore for EncryptingResourceAdapter<T> {
+    type ParamType = T;
+
+    fn load(&self, stream: &[u8]) -> Result<Self::ParamType, ASPersistenceError> {
+        let ciphertext = bincode::deserialize(stream)
+            .map_err(|source| ASPersistenceError::BincodeDe { source })?;
+        let plaintext =
+            self.cipher
+                .decrypt(&ciphertext)
+                .map_err(|err| ASPersistenceError::OtherLoad {
+                    inner: Box::new(err),
+                })?;
+        bincode::deserialize(&plaintext).map_err(|source| ASPersistenceError::BincodeDe { source })
+    }
+
+    fn store(&mut self, param: &Self::ParamType) -> Result<Vec<u8>, ASPersistenceError> {
+        let plaintext = bincode::serialize(param)
+            .map_err(|source| ASPersistenceError::BincodeSer { source })?;
+        let ciphertext =
+            self.cipher
+                .encrypt(&plaintext)
+                .map_err(|err| ASPersistenceError::OtherStore {
+                    inner: Box::new(err),
+                })?;
+        bincode::serialize(&ciphertext).map_err(|source| ASPersistenceError::BincodeSer { source })
+    }
+}
+
```

Comment:
Why are we moving this out of persistence.rs?  To me it makes sense to keep it where it was.

Reply:
`EncryptingResourceAdapter` is now not only used by `persistence.rs`. In `create_stores` in `lib.rs`, an `EncryptingResourceAdapter` is constructed to create both the persistence storage (`AtomicKeystoreStorage`) and the `AssetsStore`, so I moved it to here as an upper-level struct.